client_compat.go

```diff
@@ -0,0 +1,101 @@
+package dns
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"strings"
+	"time"
+)
+
+// ExchangeContext performs a synchronous UDP query, like Exchange. It
+// additionally obeys deadlines from the passed Context.
+func ExchangeContext(ctx context.Context, m *Msg, a string) (r *Msg, err error) {
+	println("dns: ExchangeContext: this function is deprecated")
+	client := Client{Net: "udp"}
+	r, _, err = client.ExchangeContext(ctx, m, a)
+	// ignorint rtt to leave the original ExchangeContext API unchanged, but
+	// this function will go away
+	return r, err
+}
+
+// ExchangeConn performs a synchronous query. It sends the message m via the connection
+// c and waits for a reply. The connection c is not closed by ExchangeConn.
+// This function is going away, but can easily be mimicked:
+//
+//	co := &dns.Conn{Conn: c} // c is your net.Conn
+//	co.WriteMsg(m)
+//	in, _  := co.ReadMsg()
+//	co.Close()
+//
+func ExchangeConn(c net.Conn, m *Msg) (r *Msg, err error) {
+	println("dns: ExchangeConn: this function is deprecated")
+	co := new(Conn)
+	co.Conn = c
+	if err = co.WriteMsg(m); err != nil {
+		return nil, err
+	}
+	r, err = co.ReadMsg()
+	if err == nil && r.Id != m.Id {
+		err = ErrId
+	}
+	return r, err
+}
+
+// DialTimeout acts like Dial but takes a timeout.
+func DialTimeout(network, address string, timeout time.Duration) (conn *Conn, err error) {
+	println("dns: DialTimeout: this function is deprecated")
+
+	client := Client{Net: "udp"}
+	conn, err = client.DialWithDialer(&net.Dialer{Timeout: timeout}, address)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+// DialWithTLS connects to the address on the named network with TLS.
+func DialWithTLS(network, address string, tlsConfig *tls.Config) (conn *Conn, err error) {
+	println("dns: DialWithTLS: this function is deprecated")
```

Comment:
don't think we can (or should) put these warnings here though...

Reply:
I am not aware of a blessed way to generate warnings in Go, so I just did what already done in `ExchangeConn`. According to the documentation, `println` is unstable API and may disappear.

On the "whether we should", I think it's useful to notify the user of the risk of their code breaking with future versions of the library. While it may be annoying to read on stdout/logs, it is a clear indicator of a bigger risk.

I don't have strong opinions on this, I can remove them if you think this is wrong.

Reply:
Yes, remove these println's; nothing is deprecated here, we adding new stuff.

Reply:
will do